clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

```diff
@@ -567,6 +567,44 @@ public void testMetadataTopicExpiry() throws Exception {
         assertTrue(future.isDone(), "Request should be completed");
     }
 
+    @Test
+    public void senderThreadShouldNotGetStuckWhenThrottledAndAddingPartitionsToTxn() {
+        // We want MockClient#poll() to advance time so that eventually the backoff expires.
+        client.advanceTimeDuringPoll(true);
+
+        ProducerIdAndEpoch producerIdAndEpoch = new ProducerIdAndEpoch(123456L, (short) 0);
+        apiVersions.update("0", NodeApiVersions.create(ApiKeys.INIT_PRODUCER_ID.id, (short) 0, (short) 3));
+        TransactionManager txnManager = new TransactionManager(logContext, "testUnresolvedSeq", 60000, 100, apiVersions);
+
+        setupWithTransactionState(txnManager);
+        doInitTransactions(txnManager, producerIdAndEpoch);
+
+        int throttleTimeMs = 1000;
+        long startTime = time.milliseconds();
+        Node nodeToThrottle = metadata.fetch().nodeById(0);
+        client.throttle(nodeToThrottle, throttleTimeMs);
+
+        // Verify node is throttled a little bit. In real-life Apache Kafka, we observe that this can happen
+        // as done above by throttling or with a disconnect / backoff.
+        long currentPollDelay = client.pollDelayMs(nodeToThrottle, startTime);
+        assertTrue(currentPollDelay > 0);
+        assertTrue(currentPollDelay <= throttleTimeMs);
```

Comment:
nit: I wonder if we could just assert that currentPollDelay equals to throttleTimeMs. The time does not seem to advance before pollDelayMs is called. Would it work?